core/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * // add the tracing handlers
+ * observationRegistry.observationConfig().observationHandler(
+ *         new FirstMatchingCompositeObservationHandler(tracingHandlers));
+ *
+ * // add the decorator
+ * Server.builder()
+ *       .decorator(MicrometerObservationService.newDecorator(registry))
+ * ...
+ * }</pre>
+ */
+@UnstableApi
+public final class MicrometerObservationService extends SimpleDecoratingHttpService {
+
+    /**
+     * Creates a new micrometer observation integrated {@link HttpService} decorator using the
+     * specified {@link ObservationRegistry} instance.
+     */
+    public static Function<? super HttpService, MicrometerObservationService>
+    newDecorator(ObservationRegistry observationRegistry) {
+        return service -> new MicrometerObservationService(service, observationRegistry, null);
+    }
+
+    /**
+     * Creates a new micrometer observation integrated {@link HttpService} decorator using the
+     * specified {@link ObservationRegistry} and {@link ObservationConvention}.
+     */
+    public static Function<? super HttpService, MicrometerObservationService>
+    newDecorator(ObservationRegistry observationRegistry,
+                 ObservationConvention<HttpServerContext> observationConvention) {
+        return service -> new MicrometerObservationService(
+                service, observationRegistry, requireNonNull(observationConvention, "observationConvention"));
+    }
+
+    private final ObservationRegistry observationRegistry;
+    @Nullable
+    private final ObservationConvention<HttpServerContext> observationConvention;
+
+    private MicrometerObservationService(
+            HttpService delegate, ObservationRegistry observationRegistry,
+            @Nullable ObservationConvention<HttpServerContext> observationConvention) {
+        super(delegate);
+        this.observationRegistry = requireNonNull(observationRegistry, "observationRegistry");
+        this.observationConvention = observationConvention;
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        if (!ctx.config().transientServiceOptions().contains(TransientServiceOption.WITH_TRACING) ||
+            !ctx.config().transientServiceOptions().contains(
+                    TransientServiceOption.WITH_METRIC_COLLECTION)) {
+            return unwrap().serve(ctx, req);
+        }
+
+        final HttpServerContext httpServerContext = new HttpServerContext(ctx, req);
+        final Observation observation = HttpServerObservationDocumentation.OBSERVATION.observation(
+                observationConvention, DefaultServiceObservationConvention.INSTANCE,
+                () -> httpServerContext, observationRegistry).start();
+
+        final RequestContextExtension ctxExtension = ctx.as(RequestContextExtension.class);
+        if (!observationRegistry.isNoop() && !observation.isNoop() && ctxExtension != null) {
```

Comment:
Ditto. Should we skip observation if `observationRegistry.isNoop() == true || observation.isNoop() == true`?

Reply:
Yes, we can skip any instrumentation if registry or observation are noop